cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,213 @@
+  private static final Logger LOG = LoggerFactory.getLogger(ZKRouteStore.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_INTEGER_TYPE = new TypeToken<Map<String, Integer>>() { }.getType();
+  private static final int ZK_TIMEOUT_SECS = 5;
+
+  private final ZKClient zkClient;
+  private final ConcurrentMap<ProgramId, SettableFuture<RouteConfig>> routeConfigMap;
+
+  @Inject
+  public ZKRouteStore(ZKClient zkClient) {
+    this.zkClient = zkClient;
+    this.routeConfigMap = Maps.newConcurrentMap();
+  }
+
+  @Override
+  public void store(final ProgramId serviceId, final RouteConfig routeConfig) {
+    Supplier<RouteConfig> supplier = Suppliers.ofInstance(routeConfig);
+    final SettableFuture<RouteConfig> settableFuture = SettableFuture.create();
+    Futures.addCallback(ZKExtOperations.createOrSet(zkClient, getZKPath(serviceId), supplier, ROUTE_CONFIG_CODEC, 10),
+                        new FutureCallback<RouteConfig>() {
+                          @Override
+                          public void onSuccess(RouteConfig result) {
+                            settableFuture.set(result);
+                            routeConfigMap.put(serviceId, settableFuture);
+                          }
+
+                          @Override
+                          public void onFailure(Throwable t) {
+                            settableFuture.setException(t);
+                          }
+                        });
+    try {
+      settableFuture.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException ex) {
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public void delete(final ProgramId serviceId) throws NotFoundException {
+    OperationFuture<String> future = zkClient.delete(getZKPath(serviceId));
+    final SettableFuture<String> settableFuture = SettableFuture.create();
+    Futures.addCallback(future, new FutureCallback<String>() {
+      @Override
+      public void onSuccess(String result) {
+        routeConfigMap.remove(serviceId);
+        settableFuture.set(result);
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        settableFuture.setException(t);
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+
+    try {
+      settableFuture.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException ex) {
+      if (ex.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public RouteConfig fetch(final ProgramId serviceId) throws NotFoundException {
+    if (routeConfigMap.putIfAbsent(serviceId, SettableFuture.<RouteConfig>create()) != null) {
+      return getConfig(serviceId, routeConfigMap.get(serviceId));
```

Comment:
Use the future returned by `routeConfigMap.putIfAbsent(serviceId, SettableFuture.<RouteConfig>create())` here, otherwise it leads to a race condition.